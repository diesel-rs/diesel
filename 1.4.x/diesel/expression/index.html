<!DOCTYPE html><html lang="en"><head><meta charset="utf-8"><meta name="viewport" content="width=device-width, initial-scale=1.0"><meta name="generator" content="rustdoc"><meta name="description" content="AST types representing various typed SQL expressions."><meta name="keywords" content="rust, rustlang, rust-lang, expression"><title>diesel::expression - Rust</title><link rel="stylesheet" type="text/css" href="../../normalize.css"><link rel="stylesheet" type="text/css" href="../../rustdoc.css" id="mainThemeStyle"><link rel="stylesheet" type="text/css" href="../../light.css"  id="themeStyle"><link rel="stylesheet" type="text/css" href="../../dark.css" disabled ><link rel="stylesheet" type="text/css" href="../../ayu.css" disabled ><script id="default-settings" ></script><script src="../../storage.js"></script><script src="../../crates.js"></script><noscript><link rel="stylesheet" href="../../noscript.css"></noscript><link rel="icon" type="image/svg+xml" href="../../favicon.svg"><link rel="alternate icon" type="image/png" href="../../favicon-16x16.png"><link rel="alternate icon" type="image/png" href="../../favicon-32x32.png"><style type="text/css">#crate-search{background-image:url("../../down-arrow.svg");}</style></head><body class="rustdoc mod"><!--[if lte IE 11]><div class="warning">This old browser is unsupported and will most likely display funky things.</div><![endif]--><nav class="sidebar"><div class="sidebar-menu" role="button">&#9776;</div><a href='../../diesel/index.html'><div class='logo-container rust-logo'><img src='../../rust-logo.png' alt='logo'></div></a><h2 class="location">Module expression</h2><div class="sidebar-elems"><div class="block items"><ul><li><a href="#structs">Structs</a></li><li><a href="#traits">Traits</a></li></ul></div><div id="sidebar-vars" data-name="expression" data-ty="mod" data-relpath="./"></div><script defer src="./sidebar-items.js"></script></div></nav><div class="theme-picker"><button id="theme-picker" aria-label="Pick another theme!" aria-haspopup="menu" title="themes"><img width="18" height="18" alt="Pick another theme!" src="../../brush.svg"></button><div id="theme-choices" role="menu"></div></div><nav class="sub"><form class="search-form"><div class="search-container"><div><select id="crate-search"><option value="All crates">All crates</option></select><input class="search-input"name="search" disabled autocomplete="off" spellcheck="false" placeholder="Click or press ‘S’ to search, ‘?’ for more options…" type="search"></div><button type="button" id="help-button" title="help">?</button><a id="settings-menu" href="../../settings.html" title="settings"><img width="18" height="18" alt="Change settings" src="../../wheel.svg"></a></div></form></nav><section id="main" class="content"><h1 class="fqn"><span class="in-band">Module <a href="../index.html">diesel</a>::<wbr><a class="mod" href="#">expression</a><button id="copy-path" onclick="copy_path(this)" title="copy path"><img src="../../clipboard.svg" width="19" height="18" alt="Copy item import" title="Copy item import to clipboard"></button></span><span class="out-of-band"><span id="render-detail"><a id="toggle-all-docs" href="javascript:void(0)" title="collapse all docs">[<span class="inner">&#x2212;</span>]</a></span><a class="srclink" href="../../src/diesel/expression/mod.rs.html#1-369" title="goto source code">[src]</a></span></h1><details class="rustdoc-toggle top-doc" open><summary class="hideme"><span>Expand description</span></summary><div class="docblock"><p>AST types representing various typed SQL expressions.</p>
<p>Almost all types implement either <a href="trait.Expression.html"><code>Expression</code></a> or
<a href="trait.AsExpression.html"><code>AsExpression</code></a>.</p>
<p>The most common expression to work with is a
<a href="../query_source/trait.Column.html"><code>Column</code></a>. There are various methods
that you can call on these, found in
<a href="../expression_methods"><code>expression_methods</code></a>.</p>
<p>You can also use numeric operators such as <code>+</code> on expressions of the
appropriate type.</p>
<p>Any primitive which implements <a href="../serialize/trait.ToSql.html"><code>ToSql</code></a> will
also implement <a href="trait.AsExpression.html"><code>AsExpression</code></a>, allowing it to be
used as an argument to any of the methods described here.</p>
</div></details><h2 id="structs" class="section-header"><a href="#structs">Structs</a></h2>
<div class="item-table"><div class="item-left module-item"><a class="struct" href="struct.SqlLiteral.html" title="diesel::expression::SqlLiteral struct">SqlLiteral</a></div><div class="item-right docblock-short"><p>Returned by the <a href="../dsl/fn.sql.html"><code>sql()</code></a> function.</p>
</div><div class="item-left module-item"><a class="struct" href="struct.UncheckedBind.html" title="diesel::expression::UncheckedBind struct">UncheckedBind</a></div><div class="item-right docblock-short"><p>Returned by the <a href="struct.SqlLiteral.html#method.bind" title="SqlLiteral::bind()"><code>SqlLiteral::bind()</code></a> method when binding a value to a fragment of SQL.</p>
</div></div><h2 id="traits" class="section-header"><a href="#traits">Traits</a></h2>
<div class="item-table"><div class="item-left module-item"><a class="trait" href="trait.AppearsOnTable.html" title="diesel::expression::AppearsOnTable trait">AppearsOnTable</a></div><div class="item-right docblock-short"><p>Indicates that all elements of an expression are valid given a from clause.</p>
</div><div class="item-left module-item"><a class="trait" href="trait.AsExpression.html" title="diesel::expression::AsExpression trait">AsExpression</a></div><div class="item-right docblock-short"><p>Converts a type to its representation for use in Diesel’s query builder.</p>
</div><div class="item-left module-item"><a class="trait" href="trait.AsExpressionList.html" title="diesel::expression::AsExpressionList trait">AsExpressionList</a></div><div class="item-right docblock-short"><p>Converts a tuple of values into a tuple of Diesel expressions.</p>
</div><div class="item-left module-item"><a class="trait" href="trait.BoxableExpression.html" title="diesel::expression::BoxableExpression trait">BoxableExpression</a></div><div class="item-right docblock-short"><p>Helper trait used when boxing expressions.</p>
</div><div class="item-left module-item"><a class="trait" href="trait.Expression.html" title="diesel::expression::Expression trait">Expression</a></div><div class="item-right docblock-short"><p>Represents a typed fragment of SQL.</p>
</div><div class="item-left module-item"><a class="trait" href="trait.IntoSql.html" title="diesel::expression::IntoSql trait">IntoSql</a></div><div class="item-right docblock-short"><p>Converts a type to its representation for use in Diesel’s query builder.</p>
</div><div class="item-left module-item"><a class="trait" href="trait.NonAggregate.html" title="diesel::expression::NonAggregate trait">NonAggregate</a></div><div class="item-right docblock-short"><p>Marker trait to indicate that an expression does not include any aggregate
functions.</p>
</div><div class="item-left module-item"><a class="trait" href="trait.SelectableExpression.html" title="diesel::expression::SelectableExpression trait">SelectableExpression</a></div><div class="item-right docblock-short"><p>Indicates that an expression can be selected from a source.</p>
</div></div></section><section id="search" class="content hidden"></section><div id="rustdoc-vars" data-root-path="../../" data-current-crate="diesel" data-search-index-js="../../search-index.js" data-search-js="../../search.js"></div>
    <script src="../../main.js"></script>
</body></html>