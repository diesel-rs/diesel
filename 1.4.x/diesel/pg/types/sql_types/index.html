<!DOCTYPE html><html lang="en"><head><meta charset="utf-8"><meta name="viewport" content="width=device-width, initial-scale=1.0"><meta name="generator" content="rustdoc"><meta name="description" content="PostgreSQL specific SQL types"><meta name="keywords" content="rust, rustlang, rust-lang, sql_types"><title>diesel::pg::types::sql_types - Rust</title><link rel="stylesheet" type="text/css" href="../../../../normalize.css"><link rel="stylesheet" type="text/css" href="../../../../rustdoc.css" id="mainThemeStyle"><link rel="stylesheet" type="text/css" href="../../../../light.css"  id="themeStyle"><link rel="stylesheet" type="text/css" href="../../../../dark.css" disabled ><link rel="stylesheet" type="text/css" href="../../../../ayu.css" disabled ><script id="default-settings" ></script><script src="../../../../storage.js"></script><script src="../../../../crates.js"></script><noscript><link rel="stylesheet" href="../../../../noscript.css"></noscript><link rel="icon" type="image/svg+xml" href="../../../../favicon.svg"><link rel="alternate icon" type="image/png" href="../../../../favicon-16x16.png"><link rel="alternate icon" type="image/png" href="../../../../favicon-32x32.png"><style type="text/css">#crate-search{background-image:url("../../../../down-arrow.svg");}</style></head><body class="rustdoc mod"><!--[if lte IE 11]><div class="warning">This old browser is unsupported and will most likely display funky things.</div><![endif]--><nav class="sidebar"><div class="sidebar-menu" role="button">&#9776;</div><a href='../../../../diesel/index.html'><div class='logo-container rust-logo'><img src='../../../../rust-logo.png' alt='logo'></div></a><h2 class="location">Module sql_types</h2><div class="sidebar-elems"><div class="block items"><ul><li><a href="#structs">Structs</a></li><li><a href="#types">Type Definitions</a></li></ul></div><div id="sidebar-vars" data-name="sql_types" data-ty="mod" data-relpath="./"></div><script defer src="./sidebar-items.js"></script></div></nav><div class="theme-picker"><button id="theme-picker" aria-label="Pick another theme!" aria-haspopup="menu" title="themes"><img width="18" height="18" alt="Pick another theme!" src="../../../../brush.svg"></button><div id="theme-choices" role="menu"></div></div><nav class="sub"><form class="search-form"><div class="search-container"><div><select id="crate-search"><option value="All crates">All crates</option></select><input class="search-input"name="search" disabled autocomplete="off" spellcheck="false" placeholder="Click or press ‘S’ to search, ‘?’ for more options…" type="search"></div><button type="button" id="help-button" title="help">?</button><a id="settings-menu" href="../../../../settings.html" title="settings"><img width="18" height="18" alt="Change settings" src="../../../../wheel.svg"></a></div></form></nav><section id="main" class="content"><h1 class="fqn"><span class="in-band">Module <a href="../../../index.html">diesel</a>::<wbr><a href="../../index.html">pg</a>::<wbr><a href="../index.html">types</a>::<wbr><a class="mod" href="#">sql_types</a><button id="copy-path" onclick="copy_path(this)" title="copy path"><img src="../../../../clipboard.svg" width="19" height="18" alt="Copy item import" title="Copy item import to clipboard"></button></span><span class="out-of-band"><span id="render-detail"><a id="toggle-all-docs" href="javascript:void(0)" title="collapse all docs">[<span class="inner">&#x2212;</span>]</a></span><a class="srclink" href="../../../../src/diesel/pg/types/mod.rs.html#27-493" title="goto source code">[src]</a></span></h1><details class="rustdoc-toggle top-doc" open><summary class="hideme"><span>Expand description</span></summary><div class="docblock"><p>PostgreSQL specific SQL types</p>
<p>Note: All types in this module can be accessed through <code>diesel::sql_types</code></p>
</div></details><h2 id="structs" class="section-header"><a href="#structs">Structs</a></h2>
<div class="item-table"><div class="item-left module-item"><a class="struct" href="struct.Array.html" title="diesel::pg::types::sql_types::Array struct">Array</a></div><div class="item-right docblock-short"><p>The <code>Array</code> SQL type.</p>
</div><div class="item-left module-item"><a class="struct" href="struct.Cidr.html" title="diesel::pg::types::sql_types::Cidr struct">Cidr</a></div><div class="item-right docblock-short"><p>The <a href="https://www.postgresql.org/docs/9.6/static/datatype-net-types.html"><code>CIDR</code></a> SQL type. This type can only be used with <code>feature = &quot;network-address&quot;</code></p>
</div><div class="item-left module-item"><a class="struct" href="struct.Inet.html" title="diesel::pg::types::sql_types::Inet struct">Inet</a></div><div class="item-right docblock-short"><p>The <a href="https://www.postgresql.org/docs/9.6/static/datatype-net-types.html"><code>INET</code></a> SQL type. This type can only be used with <code>feature = &quot;network-address&quot;</code></p>
</div><div class="item-left module-item"><a class="struct" href="struct.Json.html" title="diesel::pg::types::sql_types::Json struct">Json</a></div><div class="item-right docblock-short"><p>The JSON SQL type.  This type can only be used with <code>feature = &quot;serde_json&quot;</code></p>
</div><div class="item-left module-item"><a class="struct" href="struct.Jsonb.html" title="diesel::pg::types::sql_types::Jsonb struct">Jsonb</a></div><div class="item-right docblock-short"><p>The <code>jsonb</code> SQL type.  This type can only be used with <code>feature = &quot;serde_json&quot;</code></p>
</div><div class="item-left module-item"><a class="struct" href="struct.MacAddr.html" title="diesel::pg::types::sql_types::MacAddr struct">MacAddr</a></div><div class="item-right docblock-short"><p>The <a href="https://www.postgresql.org/docs/9.6/static/datatype-net-types.html"><code>MACADDR</code></a> SQL type. This type can only be used with <code>feature = &quot;network-address&quot;</code></p>
</div><div class="item-left module-item"><a class="struct" href="struct.Money.html" title="diesel::pg::types::sql_types::Money struct">Money</a></div><div class="item-right docblock-short"><p>The PostgreSQL <a href="https://www.postgresql.org/docs/9.1/static/datatype-money.html">Money</a> type.</p>
</div><div class="item-left module-item"><a class="struct" href="struct.Oid.html" title="diesel::pg::types::sql_types::Oid struct">Oid</a></div><div class="item-right docblock-short"><p>The <code>OID</code> SQL type. This is a PostgreSQL specific type.</p>
</div><div class="item-left module-item"><a class="struct" href="struct.Range.html" title="diesel::pg::types::sql_types::Range struct">Range</a></div><div class="item-right docblock-short"><p>The <code>Range</code> SQL type.</p>
</div><div class="item-left module-item"><a class="struct" href="struct.Record.html" title="diesel::pg::types::sql_types::Record struct">Record</a></div><div class="item-right docblock-short"><p>The <code>Record</code> (a.k.a. tuple) SQL type.</p>
</div><div class="item-left module-item"><a class="struct" href="struct.Timestamptz.html" title="diesel::pg::types::sql_types::Timestamptz struct">Timestamptz</a></div><div class="item-right docblock-short"><p>The “timestamp with time zone” SQL type, which PostgreSQL abbreviates
to <code>timestamptz</code>.</p>
</div><div class="item-left module-item"><a class="struct" href="struct.Uuid.html" title="diesel::pg::types::sql_types::Uuid struct">Uuid</a></div><div class="item-right docblock-short"><p>The <code>UUID</code> SQL type. This type can only be used with <code>feature = &quot;uuid&quot;</code></p>
</div></div><h2 id="types" class="section-header"><a href="#types">Type Definitions</a></h2>
<div class="item-table"><div class="item-left module-item"><a class="type" href="type.BigSerial.html" title="diesel::pg::types::sql_types::BigSerial type">BigSerial</a></div><div class="item-right docblock-short"><p>Alias for <code>BigInt</code></p>
</div><div class="item-left module-item"><a class="type" href="type.Serial.html" title="diesel::pg::types::sql_types::Serial type">Serial</a></div><div class="item-right docblock-short"><p>Alias for <code>Integer</code></p>
</div><div class="item-left module-item"><a class="type" href="type.SmallSerial.html" title="diesel::pg::types::sql_types::SmallSerial type">SmallSerial</a></div><div class="item-right docblock-short"><p>Alias for <code>SmallInt</code></p>
</div></div></section><section id="search" class="content hidden"></section><div id="rustdoc-vars" data-root-path="../../../../" data-current-crate="diesel" data-search-index-js="../../../../search-index.js" data-search-js="../../../../search.js"></div>
    <script src="../../../../main.js"></script>
</body></html>