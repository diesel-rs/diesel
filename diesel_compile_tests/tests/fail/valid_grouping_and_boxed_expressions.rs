extern crate diesel;

use diesel::expression::{is_aggregate, MixedAggregates, ValidGrouping};
use diesel::pg::Pg;
use diesel::prelude::*;
use diesel::sql_types::{Integer, Nullable};

table! {
    users {
        id -> Integer,
        name -> Text,
    }
}

fn some_ungrouped_expression(
    something: bool,
) -> Box<dyn BoxableExpression<users::table, Pg, SqlType = Integer>> {
    if something {
        Box::new(5.into_sql::<Integer>())
    } else {
        Box::new(users::id)
    }
}

fn maybe_grouped<GB>(
    something: bool,
) -> Box<
    dyn BoxableExpression<
        users::table,
        Pg,
        GB,
        <users::id as ValidGrouping<GB>>::IsAggregate,
        SqlType = Integer,
    >,
>
where
    users::id: BoxableExpression<
            users::table,
            Pg,
            GB,
            <users::id as ValidGrouping<GB>>::IsAggregate,
            SqlType = Integer,
        > + ValidGrouping<GB>,
{
    if something {
        Box::new(5.into_sql::<Integer>())
    } else {
        Box::new(users::id)
    }
}

fn something_that_is_aggregate<GB>(
) -> Box<dyn BoxableExpression<users::table, Pg, GB, is_aggregate::Yes, SqlType = Nullable<Integer>>>
where
    diesel::dsl::count<users::id>: BoxableExpression<users::table, Pg, GB, is_aggregate::Yes>,
{
    Box::new(diesel::dsl::max(users::id))
}

fn main() {
    let mut conn = PgConnection::establish("connection_string").unwrap();

    // it's fine to pass this to some query without group clause
    users::table
        .select(some_ungrouped_expression(true))
        .load::<i32>(&mut conn);

    // this fails because there is explicitly no group by clause
    users::table
        .group_by(users::id)
        .select(some_ungrouped_expression(true))
        .load::<i32>(&mut conn);

    // it's fine to pass this to some query without group by clause
    // rustc should infer the correct bounds here
    users::table.select(maybe_grouped(false)).load::<i32>(&mut conn);

    // it's also fine to pass this to some query with a matching
    // group by clause
    users::table
        .group_by(users::id)
        .select(maybe_grouped(true))
        .load::<i32>(&mut conn);

    // this fails because of an incompatible group by clause
    users::table
        .group_by(users::name)
        .select(maybe_grouped(true))
        .load::<i32>(&mut conn);

    // aggregated expressions work to
    users::table
        .select(something_that_is_aggregate())
        .load::<Option<i32>>(&mut conn);

    // also with a group by clause
    users::table
        .group_by(users::name)
        .select(something_that_is_aggregate())
        .load::<Option<i32>>(&mut conn);

    // but we cannot mix a aggregated expression with an non aggregate one
    users::table
        .select((
            something_that_is_aggregate(),
            some_ungrouped_expression(false),
        ))
        .load::<(Option<i32>, i32)>(&mut conn);

    // using two potential aggregated expressions works
    users::table
        .group_by(users::id)
        .select((something_that_is_aggregate(), maybe_grouped(true)))
        .load::<(Option<i32>, i32)>(&mut conn);
}
