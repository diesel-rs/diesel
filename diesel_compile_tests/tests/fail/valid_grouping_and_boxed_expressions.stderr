warning: unused import: `MixedAggregates`
 --> $DIR/valid_grouping_and_boxed_expressions.rs:3:40
  |
3 | use diesel::expression::{is_aggregate, MixedAggregates, ValidGrouping};
  |                                        ^^^^^^^^^^^^^^^
  |
  = note: `#[warn(unused_imports)]` on by default

error[E0277]: the trait bound `dyn diesel::BoxableExpression<users::table, Pg, SqlType = diesel::sql_types::Integer>: ValidGrouping<columns::id>` is not satisfied
  --> $DIR/valid_grouping_and_boxed_expressions.rs:71:17
   |
71 |         .select(some_ungrouped_expression(true))
   |                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `ValidGrouping<columns::id>` is not implemented for `dyn diesel::BoxableExpression<users::table, Pg, SqlType = diesel::sql_types::Integer>`
   |
   = help: the following implementations were found:
             <(dyn diesel::BoxableExpression<QS, DB, GB, IsAggregate, SqlType = ST> + 'a) as ValidGrouping<GB>>
   = note: required because of the requirements on the impl of `ValidGrouping<columns::id>` for `Box<dyn diesel::BoxableExpression<users::table, Pg, SqlType = diesel::sql_types::Integer>>`
   = note: required because of the requirements on the impl of `SelectDsl<Box<dyn diesel::BoxableExpression<users::table, Pg, SqlType = diesel::sql_types::Integer>>>` for `SelectStatement<users::table, diesel::query_builder::select_clause::DefaultSelectClause, diesel::query_builder::distinct_clause::NoDistinctClause, diesel::query_builder::where_clause::NoWhereClause, diesel::query_builder::order_clause::NoOrderClause, LimitOffsetClause<NoLimitClause, NoOffsetClause>, diesel::query_builder::group_by_clause::GroupByClause<columns::id>>`

error[E0277]: the trait bound `dyn diesel::BoxableExpression<users::table, Pg, SqlType = diesel::sql_types::Integer>: ValidGrouping<columns::id>` is not satisfied
  --> $DIR/valid_grouping_and_boxed_expressions.rs:72:10
   |
72 |         .load::<i32>(&mut conn);
   |          ^^^^ the trait `ValidGrouping<columns::id>` is not implemented for `dyn diesel::BoxableExpression<users::table, Pg, SqlType = diesel::sql_types::Integer>`
   |
   = help: the following implementations were found:
             <(dyn diesel::BoxableExpression<QS, DB, GB, IsAggregate, SqlType = ST> + 'a) as ValidGrouping<GB>>
   = note: required because of the requirements on the impl of `ValidGrouping<columns::id>` for `Box<dyn diesel::BoxableExpression<users::table, Pg, SqlType = diesel::sql_types::Integer>>`
   = note: required because of the requirements on the impl of `Query` for `SelectStatement<users::table, diesel::query_builder::select_clause::SelectClause<Box<dyn diesel::BoxableExpression<users::table, Pg, SqlType = diesel::sql_types::Integer>>>, diesel::query_builder::distinct_clause::NoDistinctClause, diesel::query_builder::where_clause::NoWhereClause, diesel::query_builder::order_clause::NoOrderClause, LimitOffsetClause<NoLimitClause, NoOffsetClause>, diesel::query_builder::group_by_clause::GroupByClause<columns::id>>`
   = note: required because of the requirements on the impl of `LoadQuery<_, i32>` for `SelectStatement<users::table, diesel::query_builder::select_clause::SelectClause<Box<dyn diesel::BoxableExpression<users::table, Pg, SqlType = diesel::sql_types::Integer>>>, diesel::query_builder::distinct_clause::NoDistinctClause, diesel::query_builder::where_clause::NoWhereClause, diesel::query_builder::order_clause::NoOrderClause, LimitOffsetClause<NoLimitClause, NoOffsetClause>, diesel::query_builder::group_by_clause::GroupByClause<columns::id>>`

error[E0271]: type mismatch resolving `<columns::name as IsContainedInGroupBy<columns::id>>::Output == diesel::expression::is_contained_in_group_by::Yes`
  --> $DIR/valid_grouping_and_boxed_expressions.rs:88:17
   |
88 |         .select(maybe_grouped(true))
   |                 ^^^^^^^^^^^^^ expected struct `diesel::expression::is_contained_in_group_by::No`, found struct `diesel::expression::is_contained_in_group_by::Yes`
   |
   = note: required because of the requirements on the impl of `ValidGrouping<columns::name>` for `columns::id`

error[E0277]: the trait bound `diesel::expression::is_aggregate::Yes: MixedAggregates<diesel::expression::is_aggregate::No>` is not satisfied
   --> $DIR/valid_grouping_and_boxed_expressions.rs:104:17
    |
104 |           .select((
    |  _________________^
105 | |             something_that_is_aggregate(),
106 | |             some_ungrouped_expression(false),
107 | |         ))
    | |_________^ the trait `MixedAggregates<diesel::expression::is_aggregate::No>` is not implemented for `diesel::expression::is_aggregate::Yes`
    |
    = help: the following implementations were found:
              <diesel::expression::is_aggregate::Yes as MixedAggregates<diesel::expression::is_aggregate::Never>>
              <diesel::expression::is_aggregate::Yes as MixedAggregates<diesel::expression::is_aggregate::Yes>>
    = note: required because of the requirements on the impl of `ValidGrouping<()>` for `(Box<dyn diesel::BoxableExpression<users::table, Pg, (), diesel::expression::is_aggregate::Yes, SqlType = diesel::sql_types::Nullable<diesel::sql_types::Integer>>>, Box<dyn diesel::BoxableExpression<users::table, Pg, SqlType = diesel::sql_types::Integer>>)`
    = note: required because of the requirements on the impl of `SelectDsl<(Box<dyn diesel::BoxableExpression<users::table, Pg, (), diesel::expression::is_aggregate::Yes, SqlType = diesel::sql_types::Nullable<diesel::sql_types::Integer>>>, Box<dyn diesel::BoxableExpression<users::table, Pg, SqlType = diesel::sql_types::Integer>>)>` for `SelectStatement<users::table>`

error[E0277]: the trait bound `diesel::expression::is_aggregate::Yes: MixedAggregates<diesel::expression::is_aggregate::No>` is not satisfied
   --> $DIR/valid_grouping_and_boxed_expressions.rs:108:10
    |
108 |         .load::<(Option<i32>, i32)>(&mut conn);
    |          ^^^^ the trait `MixedAggregates<diesel::expression::is_aggregate::No>` is not implemented for `diesel::expression::is_aggregate::Yes`
    |
    = help: the following implementations were found:
              <diesel::expression::is_aggregate::Yes as MixedAggregates<diesel::expression::is_aggregate::Never>>
              <diesel::expression::is_aggregate::Yes as MixedAggregates<diesel::expression::is_aggregate::Yes>>
    = note: required because of the requirements on the impl of `ValidGrouping<()>` for `(Box<dyn diesel::BoxableExpression<users::table, Pg, (), diesel::expression::is_aggregate::Yes, SqlType = diesel::sql_types::Nullable<diesel::sql_types::Integer>>>, Box<dyn diesel::BoxableExpression<users::table, Pg, SqlType = diesel::sql_types::Integer>>)`
    = note: required because of the requirements on the impl of `Query` for `SelectStatement<users::table, diesel::query_builder::select_clause::SelectClause<(Box<dyn diesel::BoxableExpression<users::table, Pg, (), diesel::expression::is_aggregate::Yes, SqlType = diesel::sql_types::Nullable<diesel::sql_types::Integer>>>, Box<dyn diesel::BoxableExpression<users::table, Pg, SqlType = diesel::sql_types::Integer>>)>>`
    = note: required because of the requirements on the impl of `LoadQuery<_, (Option<i32>, i32)>` for `SelectStatement<users::table, diesel::query_builder::select_clause::SelectClause<(Box<dyn diesel::BoxableExpression<users::table, Pg, (), diesel::expression::is_aggregate::Yes, SqlType = diesel::sql_types::Nullable<diesel::sql_types::Integer>>>, Box<dyn diesel::BoxableExpression<users::table, Pg, SqlType = diesel::sql_types::Integer>>)>>`
